mailbox/noise.go

```diff
@@ -697,25 +680,47 @@ func (b *Machine) GenActTwo() ([ActTwoSize]byte, error) {
 // the initiator. A successful processing of this packet presents the initiator
 // with the responder's public key, and sets us up for the final leg of the
 // triple diffie hellman.
-func (b *Machine) RecvActTwo(actTwo [ActTwoSize]byte) error {
+func (b *Machine) RecvActTwo(reader io.Reader) error {
 	var (
-		err error
-		e   [33]byte
-		s   [33 + 16]byte
-		p   [ActTwoPayloadSize + 16]byte
+		err     error
+		version [1]byte
+		e       [33]byte
+		s       [33 + 16]byte
 	)
 
+	// Read the version byte.
+	if _, err := io.ReadFull(reader, version[:]); err != nil {
+		return err
+	}
+
 	// If the handshake version is unknown, then the handshake fails
 	// immediately.
-	if actTwo[0] != HandshakeVersion {
+	if version[0] != HandshakeVersion {
 		return fmt.Errorf("act two: invalid handshake version: %v, "+
-			"only %v is valid, msg=%x", actTwo[0], HandshakeVersion,
-			actTwo[:])
+			"only %v is valid", version[0], HandshakeVersion)
+	}
+
+	// Read the ephemeral key.
+	if _, err := io.ReadFull(reader, e[:]); err != nil {
+		return err
+	}
+
+	// Read the static key.
+	if _, err := io.ReadFull(reader, s[:]); err != nil {
+		return err
 	}
 
-	copy(e[:], actTwo[1:34])
-	copy(s[:], actTwo[34:83])
-	copy(p[:], actTwo[83:])
+	// Read the payload length key.
+	var payloadLenBytes [2]byte
+	if _, err := io.ReadFull(reader, payloadLenBytes[:]); err != nil {
+		return err
+	}
+
+	authPayloadLen := binary.BigEndian.Uint16(payloadLenBytes[:])
+	p := make([]byte, authPayloadLen)
```

Comment:
If we're going with `uint32` we should probably still check the length here to avoid allocating up to 4GB of memory if the remote party is malicious. Perhaps we should limit the size to 1MB for now and error out if a larger length is sent here?

Reply:
I think a `uint16` is more than enough: why would macaroons ever get that large?

Reply:
thoughts @guggero?